### DISCUSSION TITLE
First time writing instrumentation: Muzzle help

### DISCUSSION THREAD

**Discussion by robododge** [open-telemetry/opentelemetry-java-instrumentation#2862] [Q&A] [answered]
I'm trying to implement the Jetty httpclient.   I started with the Jetty-client 9.0 series with hopes to come back to the earlier Jetty httpclient versions after.  I have read the [Muzzle doc](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/muzzle.md) back and forth.   Getting muzzle error only at runtime.

```
opentelemetry.auto.trace 2021-04-23 17:04:13:858 -0500] [main] WARN muzzleMatcher - Instrumentation skipped, mismatched references were found: jetty-httpclient-9 -- io.opentelemetry.javaagent.instrumentation.jettyhttpclient.v9_0.JettyHttpClient9InstrumentationModule on jdk.internal.loader.ClassLoaders$AppClassLoader@707f7052
[opentelemetry.auto.trace 2021-04-23 17:04:13:867 -0500] [main] WARN muzzleMatcher - -- io.opentelemetry.javaagent.instrumentation.jettyhttpclient.v9_0.JettyHttpClient9InstrumentationModule$JettyHttpClient9Instrumentation:0 Missing class io.opentelemetry.javaagent.tooling.TypeInstrumentation
[opentelemetry.auto.trace 2021-04-23 17:04:13:867 -0500] [main] WARN muzzleMatcher - -- io.opentelemetry.javaagent.instrumentation.jettyhttpclient.v9_0.JettyHttpClient9InstrumentationModule$JettyHttpClient9Instrumentation:48 Missing class net.bytebuddy.matcher.ElementMatcher$Junction$AbstractBase
[opentelemetry.auto.trace 2021-04-23 17:04:13:867 -0500] [main] WARN muzzleMatcher - -- io.opentelemetry.javaagent.instrumentation.jettyhttpclient.v9_0.JettyHttpClient9InstrumentationModule:0 Missing class io.opentelemetry.javaagent.tooling.InstrumentationModule
[opentelemetry.auto.trace 2021-04-23 17:04:13:868 -0500] [main] WARN muzzleMatcher - -- io.opentelemetry.javaagent.instrumentation.jettyhttpclient.v9_0.JettyHttpClient9InstrumentationModule$JettyHttpClient9Instrumentation:53 Missing class net.bytebuddy.matcher.ElementMatchers
[opentelemetry.auto.trace 2021-04-23 17:04:13:868 -0500] [main] WARN muzzleMatcher - -- io.opentelemetry.javaagent.instrumentation.jettyhttpclient.v9_0.JettyHttpClient9InstrumentationModule$JettyHttpClient9Instrumentation:61 Missing class net.bytebuddy.matcher.ElementMatcher
[opentelemetry.auto.trace 2021-04-23 17:04:13:868 -0500] [main] WARN muzzleMatcher - -- io.opentelemetry.javaagent.instrumentation.jettyhttpclient.v9_0.JettyHttpClient9InstrumentationModule$JettyHttpClient9Instrumentation:53 Missing class net.bytebuddy.matcher.ElementMatcher$Junction
[opentelemetry.auto.trace 2021-04-23 17:04:13:868 -0500] [main] WARN muzzleMatcher - -- io.opentelemetry.javaagent.instrumentation.jettyhttpclient.v9_0.JettyHttpClient9InstrumentationModule$JettyHttpClient9Instrumentation:48 Missing class io.opentelemetry.javaagent.tooling.bytebuddy.matcher.ClassLo

```

```
muzzle {
  pass {
    group = "org.eclipse.jetty"
    module = 'jetty-client'
    versions = "[9.0,11)"
  }
}
```

It looks like at runtime muzzle isn't allowing my new instrumentation for jetty httpclient to auto-instrument because of missing classes?   But the classes it calls out as being missing are core otel java instr classes.  

I designed my instrumentation module following the [Module Guide](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/writing-instrumentation-module.md)  I'm building on Jdk 15 then creating a full -all.jar `gradle assemble`.    Then, I have a simple main program with a Jetty client that loops.

What am I doing wrong?

**Comment by trask** [comment, marked as answer]
hey @robododge!

Based on the exception, my first guess is that you have a reference from one of your Advice classes to your InstrumentationModule.

See https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/writing-instrumentation-module.md#advice-classes, in particular:

> Advice classes are not really "classes", they're raw pieces of code that will be pasted directly into
the instrumented library class files. You should not treat them as ordinary, plain Java classes -
unfortunately many standard practices do not apply to them:
>
> * they MUST only contain static methods;
> * they MUST NOT contain any state (fields) whatsoever - static constants included! Only the advice
  methods' content is copied to the instrumented code, the constants are not;
> * reusing code by extracting a common method and/or parent class will most likely not work properly:
  instead you can create additional helper classes to store any reusable code;
> * they SHOULD NOT contain any methods other than `@Advice`-annotated method.

If this doesn't get you unstuck, feel free to send a draft PR to this repo with what you have, and we can help you identify the issue.

**Reply by robododge** [reply]
Thank you @trask,    That was definitely it!    I had a static logger instance from the InstrumentationModule referenced in my Advice class😃    Up and running now!

**Reply by mateuszrzeszutek** [reply]
Ive added some extra clarification to the doc: https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/2871
